Modules/hmacmodule.c

```diff
@@ -0,0 +1,1713 @@
+{
+    const char *utf8name = PyUnicode_AsUTF8(name);
+    if (utf8name == NULL) {
+        goto error;
+    }
+    if (find_hash_info_by_utf8name(state, utf8name, info)) {
+        return 1;
+    }
+
+    // try to find an alternative using the lowercase name
+    PyObject *lower = PyObject_CallMethodNoArgs(name, state->str_lower);
+    if (lower == NULL) {
+        goto error;
+    }
+    const char *utf8lower = PyUnicode_AsUTF8(lower);
+    if (utf8lower == NULL) {
+        Py_DECREF(lower);
+        goto error;
+    }
+    int found = find_hash_info_by_utf8name(state, utf8lower, info);
+    Py_DECREF(lower);
+    return found;
+
+error:
+    *info = NULL;
+    return -1;
+}
+
+/*
+ * Find the corresponding HMAC hash function static information.
+ *
+ * If an error occurs or if nothing can be found, this
+ * returns -1 or 0 respectively, and sets 'info' to NULL.
+ * Otherwise, this returns 1 and stores the result in 'info'.
+ *
+ * Parameters
+ *
+ *      state           The HMAC module state.
+ *      hash_info_ref   An input to hashlib.new().
+ *      info            The deduced information, if any.
+ */
+static int
+find_hash_info_impl(hmacmodule_state *state,
+                    PyObject *hash_info_ref,
+                    const py_hmac_hinfo **info)
+{
+    if (PyUnicode_Check(hash_info_ref)) {
+        return find_hash_info_by_name(state, hash_info_ref, info);
+    }
+    // NOTE(picnixz): For now, we only support named algorithms.
+    // In the future, we need to decide whether 'hashlib.openssl_md5'
+    // would make sense as an alias to 'md5' and how to remove OpenSSL.
+    *info = NULL;
+    return 0;
+}
+
+static const py_hmac_hinfo *
+find_hash_info(hmacmodule_state *state, PyObject *hash_info_ref)
+{
+    const py_hmac_hinfo *info = NULL;
+    int rc = find_hash_info_impl(state, hash_info_ref, &info);
+    // The code below could be simplfied with only 'rc == 0' case,
+    // but we are deliberately verbose to ease future improvements.
+    if (rc < 0) {
+        return NULL;
+    }
+    if (rc == 0) {
+        PyErr_Format(state->unknown_hash_error,
+                     "unsupported hash type: %R", hash_info_ref);
+        return NULL;
+    }
+    return info;
+}
+
+/* Check that the buffer length fits on a uint32_t. */
+static inline int
+has_uint32_t_buffer_length(const Py_buffer *buffer)
+{
+#ifdef Py_HMAC_SSIZE_LARGER_THAN_UINT32
+    return buffer->len <= UINT32_MAX_AS_SSIZE_T;
+#else
+    return 1;
+#endif
+}
+
+// --- HMAC object ------------------------------------------------------------
+
+/*
+ * Use the HMAC information 'info' to populate the corresponding fields.
+ *
+ * The real 'kind' for BLAKE-2 is obtained once and depends on both static
+ * capabilities (supported compiler flags) and runtime CPUID features.
+ */
+static void
+hmac_set_hinfo(hmacmodule_state *state,
+               HMACObject *self, const py_hmac_hinfo *info)
+{
+    assert(info->display_name != NULL);
```

Comment:
Those bunch of assertions could technically be removed but at least they give us better error location reports in case of a failure.